File: src/animation.rs
use crate::art::Art;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AnsiColor {
    Reset,
    DarkGray,
    Red,
    Green,
    Yellow,
    Blue,
    Magenta,
    Cyan,
    LightGray,
    BrightRed,
    BrightGreen,
    BrightYellow,
    BrightBlue,
    BrightMagenta,
    BrightCyan,
    BrightWhite,
}

impl AnsiColor {
    pub fn write_to_buf(self, buf: &mut Vec<u8>) {
        match self {
            AnsiColor::Reset => buf.extend_from_slice(b"\x1b[0m"),
            AnsiColor::DarkGray => buf.extend_from_slice(b"\x1b[90m"),
            AnsiColor::Red => buf.extend_from_slice(b"\x1b[31m"),
            AnsiColor::Green => buf.extend_from_slice(b"\x1b[32m"),
            AnsiColor::Yellow => buf.extend_from_slice(b"\x1b[33m"),
            AnsiColor::Blue => buf.extend_from_slice(b"\x1b[34m"),
            AnsiColor::Magenta => buf.extend_from_slice(b"\x1b[35m"),
            AnsiColor::Cyan => buf.extend_from_slice(b"\x1b[36m"),
            AnsiColor::LightGray => buf.extend_from_slice(b"\x1b[37m"),
            AnsiColor::BrightRed => buf.extend_from_slice(b"\x1b[91m"),
            AnsiColor::BrightGreen => buf.extend_from_slice(b"\x1b[92m"),
            AnsiColor::BrightYellow => buf.extend_from_slice(b"\x1b[93m"),
            AnsiColor::BrightBlue => buf.extend_from_slice(b"\x1b[94m"),
            AnsiColor::BrightMagenta => buf.extend_from_slice(b"\x1b[95m"),
            AnsiColor::BrightCyan => buf.extend_from_slice(b"\x1b[96m"),
            AnsiColor::BrightWhite => buf.extend_from_slice(b"\x1b[97m"),
        }
    }
}

pub fn cell_hash(seed: u64, x: usize, y: usize) -> u64 {
    let mut z = seed ^ ((x as u64) << 32 | (y as u64));
    z = z.wrapping_add(0x9E3779B97F4A7C15);
    z = (z ^ (z >> 30)).wrapping_mul(0xBF58476D1CE4E5B9);
    z = (z ^ (z >> 27)).wrapping_mul(0x94D049BB133111EB);
    z ^ (z >> 31)
}

pub fn normalized_hash(seed: u64, x: usize, y: usize) -> f32 {
    (cell_hash(seed, x, y) as f64 / u64::MAX as f64) as f32
}

#[derive(Debug, Clone)]
pub struct AnimatedCell {
    pub x: usize,
    pub y: usize,
    pub glyph: char,
    pub reveal_threshold: f32,
    pub color_phase: u8,
}

pub struct AnimationMetadata {
    pub seed: u64,
    pub width: usize,
    pub height: usize,
    pub cells: Vec<AnimatedCell>,
    cell_indices: Vec<Option<usize>>,
}

impl AnimationMetadata {
    pub fn new(art: &Art, seed: u64) -> Self {
        let mut cells = Vec::new();
        let mut cell_indices = vec![None; art.width.saturating_mul(art.height)];
        for y in 0..art.height {
            for x in 0..art.width {
                let glyph = art.get_cell(x, y);
                if glyph == ' ' {
                    continue;
                }

                let cell_index = cells.len();
                cells.push(AnimatedCell {
                    x,
                    y,
                    glyph,
                    reveal_threshold: 0.0,
                    color_phase: 0,
                });
                cell_indices[y * art.width + x] = Some(cell_index);
            }
        }

        Self {
            seed,
            width: art.width,
            height: art.height,
            cells,
            cell_indices,
        }
    }

    /// Return animation metadata for a cell in constant time.
    pub fn cell_at(&self, x: usize, y: usize) -> Option<&AnimatedCell> {
        if x >= self.width || y >= self.height {
            return None;
        }

        let index = self
            .cell_indices
            .get(y * self.width + x)
            .copied()
            .flatten()?;
        self.cells.get(index)
    }
}

pub fn smoothstep(p: f32) -> f32 {
    let clamped = p.clamp(0.0, 1.0);
    clamped * clamped * (3.0 - 2.0 * clamped)
}

/// Compute cell projected position for any of 12 30-degree angles around the 360-degree circle:
fn angle_effective_pos(x: usize, y: usize, width: usize, height: usize, angle_idx: usize) -> i32 {
    let cx = x as f32 - (width as f32 / 2.0);
    let cy = (y as f32 * 2.0) - (height as f32); // 2:1 terminal character aspect correction

    let angle_rad = (angle_idx % 12) as f32 * (std::f32::consts::PI / 6.0);
    let cos_a = angle_rad.cos();
    let sin_a = angle_rad.sin();

    (cx * cos_a + cy * sin_a) as i32
}

/// 12-Angle Directional Sweep Animation Engine (every 30 degrees):
pub fn cell_color_at(
    cell: &AnimatedCell,
    smooth_progress: f32,
    art_width: usize,
    art_height: usize,
    no_color: bool,
    iteration: usize,
) -> Option<(char, AnsiColor)> {
    if no_color {
        return Some((cell.glyph, AnsiColor::LightGray));
    }

    let max_span = (art_width
        .saturating_add(art_height.saturating_mul(2))
        .saturating_add(30) as f32
        * 0.75) as i32;

    // Pick random 12-angle sweep directions per iteration
    let dir_1 = (cell_hash(iteration as u64 ^ 0x3000, 0, 0) % 12) as usize;
    let dir_2 = ((cell_hash(iteration as u64 ^ 0x4000, 0, 0) + 1) % 12) as usize;

    // 12 standard initramfs-compatible ANSI colors
    let initramfs_12_palette = [
        AnsiColor::DarkGray,
        AnsiColor::Red,
        AnsiColor::BrightRed,
        AnsiColor::Yellow,
        AnsiColor::BrightYellow,
        AnsiColor::Green,
        AnsiColor::BrightGreen,
        AnsiColor::Cyan,
        AnsiColor::BrightCyan,
        AnsiColor::Blue,
        AnsiColor::BrightBlue,
        AnsiColor::Magenta,
    ];

    let shift_1 = iteration.wrapping_mul(5).wrapping_add(1) % initramfs_12_palette.len();
    let mut colors_0 = [AnsiColor::Reset; 7];
    for i in 0..6 {
        colors_0[i] = initramfs_12_palette[(i + shift_1) % initramfs_12_palette.len()];
    }
    colors_0[6] = AnsiColor::BrightWhite;

    let shift_2 = iteration.wrapping_mul(5).wrapping_add(7) % initramfs_12_palette.len();
    let mut colors_1 = [AnsiColor::Reset; 7];
    for i in 0..6 {
        colors_1[i] = initramfs_12_palette
            [initramfs_12_palette.len() - 1 - (i + shift_2) % initramfs_12_palette.len()];
    }
    colors_1[6] = AnsiColor::Reset;

    if smooth_progress < 0.45 {
        // Pass 1: Random 12-angle direction sweep in
        let sub_p = smooth_progress / 0.45;
        let t = (sub_p * (max_span as f32 * 2.0 + 30.0)) as i32;

        let pos = angle_effective_pos(cell.x, cell.y, art_width, art_height, dir_1);
        let f = pos - max_span + t;
        let off = ((cell_hash(42 ^ (t as u64), cell.x, cell.y) % 15) + 1) as i32;

        for i in (0..=6).rev() {
            if f > (i * 3) + off {
                return Some((cell.glyph, colors_0[i as usize]));
            }
        }
        None // 100% empty black screen
    } else if smooth_progress < 0.55 {
        // Hold Full Logo in BrightWhite
        Some((cell.glyph, AnsiColor::BrightWhite))
    } else if smooth_progress < 0.95 {
        // Pass 2: Random 12-angle direction sweep out
        let sub_p = (smooth_progress - 0.55) / 0.40;
        let t = (sub_p * (max_span as f32 * 2.0 + 30.0)) as i32;

        let pos = angle_effective_pos(cell.x, cell.y, art_width, art_height, dir_2);
        let f = pos - max_span + t;
        let off = ((cell_hash(42 ^ (t as u64), cell.x, cell.y) % 15) + 1) as i32;

        for i in (0..=6).rev() {
            if f > (i * 3) + off {
                let col = colors_1[i as usize];
                if col == AnsiColor::Reset {
                    return None; // Disappeared into blackness
                }
                return Some((cell.glyph, col));
            }
        }
        Some((cell.glyph, AnsiColor::BrightWhite)) // Full logo before wave erases cell
    } else {
        // Hold 100% Empty Black Screen
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cell_hash_deterministic() {
        let h1 = cell_hash(42, 10, 5);
        let h2 = cell_hash(42, 10, 5);
        let h3 = cell_hash(43, 10, 5);
        assert_eq!(h1, h2);
        assert_ne!(h1, h3);
    }

    #[test]
    fn test_normalized_hash_range() {
        for seed in 0..10 {
            for x in 0..5 {
                for y in 0..5 {
                    let val = normalized_hash(seed, x, y);
                    assert!((0.0..=1.0).contains(&val));
                }
            }
        }
    }

    #[test]
    fn test_smoothstep() {
        assert_eq!(smoothstep(0.0), 0.0);
        assert_eq!(smoothstep(1.0), 1.0);
        assert_eq!(smoothstep(0.5), 0.5);
    }

    #[test]
    fn test_metadata_cell_lookup() {
        let art = Art::parse("X X\n XX").unwrap();
        let metadata = AnimationMetadata::new(&art, 42);

        assert_eq!(metadata.cell_at(0, 0).map(|cell| cell.glyph), Some('X'));
        assert!(metadata.cell_at(1, 0).is_none());
        assert_eq!(metadata.cell_at(2, 1).map(|cell| cell.glyph), Some('X'));
        assert!(metadata.cell_at(3, 0).is_none());
        assert!(metadata.cell_at(0, 2).is_none());
    }
}

File: src/visual/animation.cpp
#include "sart/visual/animation.hpp"

#include <algorithm>
#include <array>
#include <bit>
#include <cmath>
#include <limits>
#include <numbers>
#include <string_view>

namespace sart::visual {
    namespace {

        constexpr std::size_t missing_index = std::numeric_limits<std::size_t>::max();

        std::int32_t angle_effective_position(std::size_t x, std::size_t y, std::size_t width, std::size_t height,
                                              std::size_t angle_index) {
            const auto center_x = static_cast<float>(x) - static_cast<float>(width) / 2.0F;
            const auto center_y = static_cast<float>(y) * 2.0F - static_cast<float>(height);
            const auto angle = static_cast<float>(angle_index % 12) * std::numbers::pi_v<float> / 6.0F;
            return static_cast<std::int32_t>(center_x * std::cos(angle) + center_y * std::sin(angle));
        }

    } // namespace

    void append_ansi_color(std::string &output, AnsiColor color) {
        static constexpr std::array<std::string_view, 16> sequences{
            "\x1b[0m",  "\x1b[90m", "\x1b[31m", "\x1b[32m", "\x1b[33m", "\x1b[34m", "\x1b[35m", "\x1b[36m",
            "\x1b[37m", "\x1b[91m", "\x1b[92m", "\x1b[93m", "\x1b[94m", "\x1b[95m", "\x1b[96m", "\x1b[97m",
        };
        output.append(sequences.at(static_cast<std::size_t>(color)));
    }

    std::uint64_t cell_hash(std::uint64_t seed, std::size_t x, std::size_t y) noexcept {
        auto value = seed ^ ((static_cast<std::uint64_t>(x) << 32) | static_cast<std::uint64_t>(y));
        value += 0x9E3779B97F4A7C15ULL;
        value = (value ^ (value >> 30)) * 0xBF58476D1CE4E5B9ULL;
        value = (value ^ (value >> 27)) * 0x94D049BB133111EBULL;
        return value ^ (value >> 31);
    }

    float normalized_hash(std::uint64_t seed, std::size_t x, std::size_t y) noexcept {
        return static_cast<float>(static_cast<double>(cell_hash(seed, x, y)) /
                                  static_cast<double>(std::numeric_limits<std::uint64_t>::max()));
    }

    float smoothstep(float progress) noexcept {
        const auto clamped = std::clamp(progress, 0.0F, 1.0F);
        return clamped * clamped * (3.0F - 2.0F * clamped);
    }

    AnimationMetadata::AnimationMetadata(const Art &art, std::uint64_t seed)
        : seed_(seed), width_(art.width()), height_(art.height()) {
        cell_indices_.assign(width_ * height_, missing_index);
        for (std::size_t y = 0; y < height_; ++y) {
            for (std::size_t x = 0; x < width_; ++x) {
                const auto glyph = art.cell(x, y);
                if (glyph == U' ') {
                    continue;
                }
                cell_indices_[y * width_ + x] = cells_.size();
                cells_.push_back({x, y, glyph, 0.0F, 0});
            }
        }
    }

    const AnimatedCell *AnimationMetadata::cell_at(std::size_t x, std::size_t y) const noexcept {
        if (x >= width_ || y >= height_) {
            return nullptr;
        }
        const auto index = cell_indices_[y * width_ + x];
        return index == missing_index ? nullptr : &cells_[index];
    }

    std::uint64_t AnimationMetadata::seed() const noexcept { return seed_; }

    std::size_t AnimationMetadata::width() const noexcept { return width_; }

    std::size_t AnimationMetadata::height() const noexcept { return height_; }

    const std::vector<AnimatedCell> &AnimationMetadata::cells() const noexcept { return cells_; }

    std::optional<ColoredGlyph> cell_color_at(const AnimatedCell &cell, float smooth_progress, std::size_t art_width,
                                              std::size_t art_height, bool no_color, std::size_t iteration) {
        if (no_color) {
            return ColoredGlyph{cell.glyph, AnsiColor::light_gray};
        }

        const auto span_source = art_width + art_height * 2 + 30;
        const auto maximum_span = static_cast<std::int32_t>(static_cast<float>(span_source) * 0.75F);
        const auto first_direction = static_cast<std::size_t>(cell_hash(iteration ^ 0x3000, 0, 0) % 12);
        const auto second_direction = static_cast<std::size_t>((cell_hash(iteration ^ 0x4000, 0, 0) + 1) % 12);
        static constexpr std::array<AnsiColor, 12> palette{
            AnsiColor::dark_gray,     AnsiColor::red,   AnsiColor::bright_red,   AnsiColor::yellow,
            AnsiColor::bright_yellow, AnsiColor::green, AnsiColor::bright_green, AnsiColor::cyan,
            AnsiColor::bright_cyan,   AnsiColor::blue,  AnsiColor::bright_blue,  AnsiColor::magenta,
        };

        std::array<AnsiColor, 7> first_colors{};
        const auto first_shift = (iteration * 5 + 1) % palette.size();
        for (std::size_t index = 0; index < 6; ++index) {
            first_colors[index] = palette[(index + first_shift) % palette.size()];
        }
        first_colors[6] = AnsiColor::bright_white;

        std::array<AnsiColor, 7> second_colors{};
        const auto second_shift = (iteration * 5 + 7) % palette.size();
        for (std::size_t index = 0; index < 6; ++index) {
            second_colors[index] = palette[palette.size() - 1 - (index + second_shift) % palette.size()];
        }
        second_colors[6] = AnsiColor::reset;

        const auto swept_color = [&](float sub_progress, std::size_t direction, const auto &colors) {
            const auto sweep =
                static_cast<std::int32_t>(sub_progress * (static_cast<float>(maximum_span) * 2.0F + 30.0F));
            const auto position = angle_effective_position(cell.x, cell.y, art_width, art_height, direction);
            const auto front = position - maximum_span + sweep;
            const auto offset =
                static_cast<std::int32_t>((cell_hash(42 ^ static_cast<std::uint64_t>(sweep), cell.x, cell.y) % 15) + 1);
            for (std::int32_t index = 6; index >= 0; --index) {
                if (front > index * 3 + offset) {
                    return std::optional<AnsiColor>{colors[static_cast<std::size_t>(index)]};
                }
            }
            return std::optional<AnsiColor>{};
        };

        if (smooth_progress < 0.45F) {
            const auto color = swept_color(smooth_progress / 0.45F, first_direction, first_colors);
            return color ? std::optional<ColoredGlyph>{{cell.glyph, *color}} : std::nullopt;
        }
        if (smooth_progress < 0.55F) {
            return ColoredGlyph{cell.glyph, AnsiColor::bright_white};
        }
        if (smooth_progress < 0.95F) {
            const auto color = swept_color((smooth_progress - 0.55F) / 0.40F, second_direction, second_colors);
            if (!color) {
                return ColoredGlyph{cell.glyph, AnsiColor::bright_white};
            }
            if (*color == AnsiColor::reset) {
                return std::nullopt;
            }
            return ColoredGlyph{cell.glyph, *color};
        }
        return std::nullopt;
    }

} // namespace sart::visual
